src/CameraOnvif.cpp

```diff
@@ -0,0 +1,408 @@
+        reportError();
+}
+
+void CameraOnvif::setResolution(int width, int height){
+    bool valid = true;
+    for (auto & e : m_private->video_options->Options->H264->ResolutionsAvailable){
+        if (e->Width == width && e->Height == height){
+            valid = true;
+        }
+    }
+
+    _trt__GetVideoEncoderConfiguration GetVideoEncoderConfiguration;
+    _trt__GetVideoEncoderConfigurationResponse GetVideoEncoderConfigurationResponse;
+    GetVideoEncoderConfiguration.ConfigurationToken = video_conf_token;
+
+    setCredentials();
+    if (m_private->proxy_media->GetVideoEncoderConfiguration(&GetVideoEncoderConfiguration,
+        GetVideoEncoderConfigurationResponse)){
+        reportError();
+    }
+
+    _trt__SetVideoEncoderConfiguration SetVideoEncoderConfiguration;
+    _trt__SetVideoEncoderConfigurationResponse SetVideoEncoderConfigurationResponse;
+    SetVideoEncoderConfiguration.Configuration =
+    GetVideoEncoderConfigurationResponse.Configuration;
+    // SetVideoEncoderConfiguration.Configuration->Encoding = tt__VideoEncoding__H264;
+    // SetVideoEncoderConfiguration.Configuration->Quality = 6;
+    // SetVideoEncoderConfiguration.Configuration->Multicast = new tt__MulticastConfiguration();
+    // SetVideoEncoderConfiguration.ForcePersistence = true;
+    // SetVideoEncoderConfiguration.Configuration->SessionTimeout = "10";
+    // SetVideoEncoderConfiguration.Configuration->H264 = new tt__H264Configuration();
+    // SetVideoEncoderConfiguration.soap = m_private->soap;
+    // SetVideoEncoderConfiguration.Configuration->H264->soap = m_private->soap;
+
+    // if (width == 1920 and height == 1080){
+    //     SetVideoEncoderConfiguration.Configuration->H264->H264Profile =
+    //     tt__H264Profile__High;
+    // }else if (width == 1280 and height == 960){
+    //     SetVideoEncoderConfiguration.Configuration->H264->H264Profile =
+    //     tt__H264Profile__Main;
+    // }else if (width == 1280 and height == 720){
+    //     SetVideoEncoderConfiguration.Configuration->H264->H264Profile =
+    //     tt__H264Profile__Baseline;
+    // }
+    if(!valid) {
+        cout <<"You are trying to set a video resolution that is not available!"
+        << endl;
+        exit(EXIT_FAILURE);
+    }
+    // SetVideoEncoderConfiguration.Configuration->Resolution = new tt__VideoResolution();
+    // SetVideoEncoderConfiguration.Configuration->Resolution->soap = m_private->soap;
+    SetVideoEncoderConfiguration.Configuration->Resolution->Width = width;
+    SetVideoEncoderConfiguration.Configuration->Resolution->Height = height;
+    setCredentials();
+    if (m_private->proxy_media->SetVideoEncoderConfiguration(&SetVideoEncoderConfiguration,
+        SetVideoEncoderConfigurationResponse)){
+        reportError();
+    }
+}
+
+// to report an error
+void CameraOnvif::reportError(){
+  cerr << "Oops, something went wrong:" << endl;
+  soap_stream_fault(m_private->soap, cerr);
+  exit(EXIT_FAILURE);
```

Comment:
Never. EVER. Ever. _ever_ use exit() within a library.

Deciding what to do on error is not your job, it's the application integrator job. You don't know what else is running in the process.

Throw an exception on error, with a proper error message. Create for instance a `SOAPException` as subclass of `std::runtime_error` and build the error message using `soap_stream_fault` (hint: std::ostringstream).